### PR TITLE
Use server-side filtering for CLI list commands

### LINKED ISSUE #10013: Use server-side filtering for CLI list commands
This could be used to make the filtering more efficient perhaps?

### PR DESCRIPTION
**Pull request by presztak**
This PR adds support for using server-side filtering for CLI list commands. Closes #10013

### CODE REVIEW

**Comment by presztak** [comment]
I have some doubts:

1. In some cases CLI filtering supports regular expressions e.g.
`lxc list user.blah=h.*`

currently this is not supported on server side. Should we support this on server side?

2. CLI filtering supports case sensitive so for e.g.

`lxc list state=running` and `lxc list state=Running` is equivalent.

This is not supported on server side. Should we support this on server side?

3. CLI filtering supports abbreviated to the smallest unambiguous identifier for configuration items e.g.

`lxc list u.blah=abc`

Should we support this on server side or let cli to take care of it?

4. CLI allows us to filter by `ExpandedConfig` in case of instances and `Properties` in case of images without specify ExpandedConfig/Properties field. So we can do:

`lxc list u.blah=abc`

but to be able to do this on server side we need to specify ExpandedConfig/Properties field. So we need to do:

`lxc list expanded_config.u.blah=abc`

What will be the best way to support this and have backward compatibility?

**Comment by presztak** [comment]
@tomponline @stgraber any thoughts regarding my above comments?

**Comment by tomponline** [comment]
@presztak sorry for not looking at this right now we've been very busy with the 5.0 LTS release and follow up fixes. Will look ASAP, thanks! :)

File: client/lxd_images.go [hunk 1]
```diff
@@ -930,3 +951,28 @@ func (r *ProtocolLXD) ExportImage(fingerprint string, image api.ImageExportPost)
 	return op, nil
 
 }
+
+// GetServerSupportedFilters returns two lists: one with filters supported by server and second one with not supported.
+func (r *ProtocolLXD) GetServerSupportedFilters(filters []string, i interface{}) ([]string, []string) {
```
Comment (stgraber) [inline client/lxd_images.go:956]: This feels a bit odd. I think it'd probably be simpler to turn this into an internal `validateFilters(filters []string) error` which then errors out if missing the extension or if the filter syntax is invalid.

Ultimately it's going to be up to the caller of those functions to pass in valid filters and to the server to return an error when something unsupported is passed.

File: client/lxd_images.go [hunk 2]
```diff
@@ -33,6 +34,26 @@ func (r *ProtocolLXD) GetImages() ([]api.Image, error) {
 	return images, nil
 }
 
+// GetFilteredImages returns a filtered list of available images as Image structs
+func (r *ProtocolLXD) GetFilteredImages(filters []string) ([]api.Image, error) {
```
Comment (stgraber) [inline client/lxd_images.go:38]: Maybe `GetImagesWithFilter` instead?

File: client/lxd_instances.go [hunk 3]
```diff
@@ -116,6 +116,31 @@ func (r *ProtocolLXD) GetInstances(instanceType api.InstanceType) ([]api.Instanc
 	return instances, nil
 }
 
+// GetFilteredInstances returns a filtered list of instances.
+func (r *ProtocolLXD) GetFilteredInstances(instanceType api.InstanceType, filters []string) ([]api.Instance, error) {
```
Comment (stgraber) [inline client/lxd_instances.go:120]: Maybe `GetInstancesWithFilter` instead?

File: client/lxd_instances.go [hunk 4]
```diff
@@ -141,6 +166,36 @@ func (r *ProtocolLXD) GetInstancesAllProjects(instanceType api.InstanceType) ([]
 	return instances, nil
 }
 
+// GetFilteredInstancesAllProjects returns a filtered list of instances from all projects.
+func (r *ProtocolLXD) GetFilteredInstancesAllProjects(instanceType api.InstanceType, filters []string) ([]api.Instance, error) {
```
Comment (stgraber) [inline client/lxd_instances.go:170]: Maybe `GetInstancesAllProjectsWithFilter` instead?

File: client/lxd_instances.go [hunk 5]
```diff
@@ -181,6 +236,35 @@ func (r *ProtocolLXD) GetInstancesFull(instanceType api.InstanceType) ([]api.Ins
 	return instances, nil
 }
 
+// GetFilteredInstancesFull returns a filtered list of instances including snapshots, backups and state.
+func (r *ProtocolLXD) GetFilteredInstancesFull(instanceType api.InstanceType, filters []string) ([]api.InstanceFull, error) {
```
Comment (stgraber) [inline client/lxd_instances.go:240]: Maybe `GetInstancesFullWithFilter` instead?

File: client/lxd_instances.go [hunk 6]
```diff
@@ -210,6 +294,40 @@ func (r *ProtocolLXD) GetInstancesFullAllProjects(instanceType api.InstanceType)
 	return instances, nil
 }
 
+// GetFilteredInstancesFullAllProjects returns a filtered list of instances including snapshots, backups and state from all projects.
+func (r *ProtocolLXD) GetFilteredInstancesFullAllProjects(instanceType api.InstanceType, filters []string) ([]api.InstanceFull, error) {
```
Comment (stgraber) [inline client/lxd_instances.go:298]: Maybe `GetInstancesFullAllProjectsWithFilter` instead (this is getting long ;))?

File: client/lxd_storage_volumes.go [hunk 7]
```diff
@@ -50,6 +50,26 @@ func (r *ProtocolLXD) GetStoragePoolVolumes(pool string) ([]api.StorageVolume, e
 	return volumes, nil
 }
 
+// GetFilteredStoragePoolVolumes returns a filtered list of StorageVolume entries for the provided pool
+func (r *ProtocolLXD) GetFilteredStoragePoolVolumes(pool string, filters []string) ([]api.StorageVolume, error) {
```
Comment (stgraber) [inline client/lxd_storage_volumes.go:54]: Maybe `GetStoragePoolVolumesWithFilter` instead?

File: client/simplestreams_images.go [hunk 8]
```diff
@@ -21,6 +21,11 @@ func (r *ProtocolSimpleStreams) GetImages() ([]api.Image, error) {
 	return r.ssClient.ListImages()
 }
 
+// GetFilteredImages returns a filtered list of available images as Image structs
+func (r *ProtocolSimpleStreams) GetFilteredImages(filters []string) ([]api.Image, error) {
```
Comment (stgraber) [inline client/simplestreams_images.go:25]: I think this should return an error instead.

File: client/simplestreams_images.go [hunk 9]
```diff
@@ -282,3 +287,8 @@ func (r *ProtocolSimpleStreams) GetImageAliasArchitectures(imageType string, nam
 func (r *ProtocolSimpleStreams) ExportImage(fingerprint string, image api.ImageExportPost) (Operation, error) {
 	return nil, fmt.Errorf("Exporting images is not supported by the simplestreams protocol")
 }
+
+// GetServerSupportedFilters returns two lists: one with filters supported by server and second one with not supported.
+func (r *ProtocolSimpleStreams) GetServerSupportedFilters(filters []string, i interface{}) ([]string, []string) {
+	return []string{}, filters
+}
```
Comment (stgraber) [inline client/simplestreams_images.go:294]: This should go away if we move to `validateFilters` or something like that instead.

File: lxc/list.go [hunk 10]
```diff
@@ -512,41 +512,45 @@ func (c *cmdList) Run(cmd *cobra.Command, args []string) error {
...
 		}
 
-		return c.showInstances(cts, filters, columns)
+		return c.showInstances(cts, clientFilters, columns)
 	}
 
 	// Get the list of instances
 	var cts []api.Instance
 	var ctslist []api.Instance
+	serverFilters, clientFilters := d.GetServerSupportedFilters(filters, api.Instance{})
```
Comment (stgraber) [inline lxc/list.go:532]: Ah, I see why that function exists now :)

I'm still not sure it really fits with the `client` package though, so maybe easier to keep it in lxc/utils for now?

File: client/lxd_images.go (hunk 1 above)
Comment (presztak) [reply client/lxd_images.go]: I've moved this method to `lxc/utils`.

File: client/lxd_images.go (hunk 2 above)
Comment (presztak) [reply client/lxd_images.go]: Fixed.

File: client/lxd_instances.go (hunk 3 above)
Comment (presztak) [reply client/lxd_instances.go]: Fixed.

File: client/lxd_instances.go (hunk 4 above)
Comment (presztak) [reply client/lxd_instances.go]: Fixed.

File: client/lxd_instances.go (hunk 5 above)
Comment (presztak) [reply client/lxd_instances.go]: Fixed.

File: client/lxd_storage_volumes.go (hunk 7 above)
Comment (presztak) [reply client/lxd_storage_volumes.go]: Fixed.

File: client/simplestreams_images.go (hunk 9 above)
Comment (presztak) [reply client/simplestreams_images.go]: I've moved this method to `lxc/utils`.

**Comment by presztak** [comment]
Hey @stgraber,

I've fixed all yours comments. Just one thing left, I didn't rename `GetServerSupportedFilters` to `validateFilters` but I've moved it to `lxc/utils`.
To could proceed this issue I still need answers for my above questions :)

**Comment by stgraber** [comment]
> 1. In some cases CLI filtering supports regular expressions e.g.
>        `lxc list user.blah=h.*`
>        currently this is not supported on server side. Should we support this on server side?

Seems fine to support server-side.



> 2\. CLI filtering supports case sensitive so for e.g.
>         `lxc list state=running` and `lxc list state=Running` is equivalent.
>         This is not supported on server side. Should we support this on server side?

Happy to have state be case insensitive on the server side.



> 3\. CLI filtering supports abbreviated to the smallest unambiguous identifier for configuration items e.g.
>         `lxc list u.blah=abc`
>         Should we support this on server side or let cli to take care of it?

I think that's best done by the client.



> 4\. CLI allows us to filter by `ExpandedConfig` in case of instances and `Properties` in case of images without specify ExpandedConfig/Properties field. So we can do:
>         `lxc list u.blah=abc`
>         but to be able to do this on server side we need to specify ExpandedConfig/Properties field. So we need to do:
>         `lxc list expanded_config.u.blah=abc`
>         What will be the best way to support this and have backward compatibility?

Can we do the naive thing of assuming that anything with a `.` in it is going to be a filter for ExpandedConfig?
Properties should be flat so no dots in them.

**Comment by tomponline** [comment]
@presztak please can you rebase, and then it will be ready to merge if @stgraber is happy with it.

**Comment by stgraber** [comment]
Did the rebase.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Thanks!

**Pull request merged**